### ISSUE TITLE
Rule ID: S2326 false positive when implementing strict generic interface

### ISSUE THREAD

**Issue by LucasLopesr** [SonarSource/sonar-dotnet#6133]
### Description

Rule ID: S2326
I'm creating a generic interface in order to restrict the parameters used and at the same time the parameters will be used to identify the IHubContext, I believe the problem is a false positive.

### Repro steps
Interface
```c#
    public interface IHubPublisher<THub> where THub : Hub
    {
        public Task NotifyGroupAsync<TMessage>(TMessage message, INotification notification)
            where TMessage : class, IMessage;
    }
```
Implementation
```c#
    public class HubPublisher<THub> : IHubPublisher<THub> where THub : Hub
    {
        private readonly IHubContext<THub> _hubContext;
        public HubPublisher(IHubContext<THub> hubContext)
        {
            _hubContext = hubContext;
        }

        public async Task NotifyGroupAsync<TMessage>(TMessage message, Notification notification)
            where TMessage : class, IMessage
        {
           // implementation...
        }
     }
```
Decorator Pattern
```c#
    public class HubPublisherXptoDecorator<THub> : IHubPublisher<THub> where THub : Hub
    {
        private readonly IHubPublisher<THub> _decoratedPublisher;

        public HubPublisherXPTODecorator(IHubPublisher<THub> decoratedPublisher)
        {
            _decoratedPublisher = decoratedPublisher;
        }

        public async Task NotifyGroupAsync<TMessage>(TMessage message, Notification notification) where TMessage 
             : class, IMessage
        {
           //implementation decorator
        }
     }
```

### Expected behavior

I believe I shouldn't criticize the use of this approach

### Related information

* C# Plugins version
* Visual Studio version 17.3.4
* dotnet version 3.1.302
* SonarScanner for .NET version (if used)
* Operating System Windows 10

**Comment by andrei-epure-sonarsource** [comment]
Thank you for opening this issue, @LucasLopesr . Your feedback helps us improve our products. 

We will look at it as soon as possible. After we confirm the issue, we will add labels and put it in our backlog.

**Comment by pavel-mikula-sonarsource** [comment]
Hi @LucasLopesr,

This looks like a tricky case to decide. From the interface point of view, the `THub` is indeed unused. And it doesn't need to be there from an interface contract point of view. It can be on the implementation side if it's used inside `NotifyGroupAsync`

How is the interface used on the invocation side? So it would really need to be generic.

**Comment by LucasLopesr** [comment]
Sorry for the delay...

For example:

Service registrations
```c#
      services.AddScoped<IHubPublisher<XptoNotificationHub>, HubPublisher<XptoNotificationHub>>()
         .Decorate<IHubPublisher<XptoNotificationHub>, HubPublisherXptoDecorator<XptoNotificationHub>>();
      services.AddScoped<IHubPublisher<AnotherXptoNotificationHub>, HubPublisher<AnotherXptoNotificationHub>>();
```

Using:

```c#
      private readonly IHubPublisher<XptoNotificationHub> _xptoHubPublisher;
      private readonly IHubPublisher<AnotherXptoNotificationHub> _anotherHubPublisher;
      public SomeOnePublisher(IHubPublisher<XptoNotificationHub> xptoHubPublisher, IHubPublisher<AnotherXptoNotificationHub> anotherHubPublisher)
        {
            _xptoHubPublisher = xptoHubPublisher;
            _anotherHubPublisher = anotherHubPublisher
        }
        
      public async Task NotifyExecutionAsync(Guid anyId, AnyEvent anyEvent)
        {           
            // implementation
            // for example
            await _xptoHubPublisher.NotifyGroupAsync(anyEvent, new Notification(groupName, notificationMessage));
            await _anotherHubPublisher.NotifyGroupAsync(anyEvent, new Notification(groupName, notificationMessage));
        }  
```
I believe that restrictions on the interface is a way of using it, no?

**Comment by csaba-sagi-sonarsource** [comment]
Hello @LucasLopesr 

If I understand your needs correctly, you could register both implementations under the `IHubPublisher` interface and then retrieve them as an `IEnumberable<IHubPublisher>`. If that is correct, then indeed you can remove the unused type parameter `THub`.
Please take a look at the following documentation for more information: https://learn.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-7.0#service-registration-methods

**Comment by csaba-sagi-sonarsource** [comment]
@LucasLopesr

Can you confirm that you have managed to remove the unused THub parameter?

**Comment by csaba-sagi-sonarsource** [comment]
As there is no reply, I am closing the issue.
